components/modules/ble.c

```diff
@@ -0,0 +1,1062 @@
+          flags |= BLE_GATT_CHR_F_WRITE;
+        }
+        if (flags) {
+          chr->flags = flags;
+        }
+
+        lua_pop(L, 3); // pop off value, read, write
+      } else {
+        lua_getfield(L, -2, "read");
+	if (!lua_isnoneornil (L, -1)) {
+	  luaL_checkfunction (L, -1);
+	  chr->flags |= BLE_GATT_CHR_F_READ;
+	}
+
+        lua_getfield(L, -3, "write");
+	if (!lua_isnoneornil (L, -1)) {
+	  luaL_checkfunction (L, -1);
+	  chr->flags |= BLE_GATT_CHR_F_WRITE;
+	}
+
+        lua_pop(L, 3); // pop off value, read, write
+      }
+
+      if (lua_getfield(L, -1, "notify") != LUA_TNIL) {
+	  chr->flags |= BLE_GATT_CHR_F_NOTIFY;
+
+          handles[0]++;
+	  lua_pushinteger(L, handles[0]);
+          lua_setfield(L, -3, "notify");
+          chr->val_handle = &handles[handles[0]];
+      }
+      lua_pop(L, 1);
+
+      // -1 is now the characteristic again
+      chr->arg = (void *) luaL_ref(L, LUA_REGISTRYINDEX);
+      chr->access_cb = lble_access_cb;
+    }
+    lua_pop(L, 2);
+    chrs++;	// terminate the list of characteristics for this service
+  }
+  lua_pop(L, 1);
+
+  *resultp = result;
+  *handlep = handles;
+
+  return 0;
+}
+
+static int
+gatt_svr_init(lua_State *L) {
+    int rc;
+
+    // Now we have to build the gatt_svr_svcs data structure
+
+
+    struct ble_gatt_svc_def *svcs = NULL;
+    lble_build_gatt_svcs(L, &svcs, &notify_handles);
+    //free_gatt_svcs(L, gatt_svr_svcs);
+    gatt_svr_svcs = svcs;
+
+    rc = ble_gatts_count_cfg(gatt_svr_svcs);
+    if (rc != 0) {
+      return luaL_error(L, "Failed to count gatts: %d", rc);
+    }
+
+    rc = ble_gatts_add_svcs(gatt_svr_svcs);
+    if (rc != 0) {
+      return luaL_error(L, "Failed to add gatts: %d", rc);
+    }
+
+    return 0;
+}
+
+/**
+ * Logs information about a connection to the console.
+ */
+static void
+lble_print_conn_desc(struct ble_gap_conn_desc *desc)
+{
+    MODLOG_DFLT(INFO, "handle=%d our_ota_addr_type=%d our_ota_addr=",
+                desc->conn_handle, desc->our_ota_addr.type);
+    print_addr(desc->our_ota_addr.val);
+    MODLOG_DFLT(INFO, " our_id_addr_type=%d our_id_addr=",
+                desc->our_id_addr.type);
+    print_addr(desc->our_id_addr.val);
+    MODLOG_DFLT(INFO, " peer_ota_addr_type=%d peer_ota_addr=",
+                desc->peer_ota_addr.type);
+    print_addr(desc->peer_ota_addr.val);
+    MODLOG_DFLT(INFO, " peer_id_addr_type=%d peer_id_addr=",
+                desc->peer_id_addr.type);
+    print_addr(desc->peer_id_addr.val);
+    MODLOG_DFLT(INFO, " conn_itvl=%d conn_latency=%d supervision_timeout=%d "
+                "encrypted=%d authenticated=%d bonded=%d\n",
+                desc->conn_itvl, desc->conn_latency,
+                desc->supervision_timeout,
+                desc->sec_state.encrypted,
+                desc->sec_state.authenticated,
+                desc->sec_state.bonded);
+}
+
+static int
+lble_sys_init(lua_State *L) {
+  int rc = nvs_flash_init();
```

Comment:
Hang on, won't this interfere with our existing NVS setup over in `user_main.c`? If anything, I'd say this logic should be moved over there if we need this sort of recovery.

Reply:
This is a bit worrying. I wouldn't have thought that I would have added that code without needing to -- I'll do some testing without it.